File: codepilot/feishu_worker.mjs
import * as Lark from '@larksuiteoapi/node-sdk';
import { spawnSync } from 'node:child_process';

const appId = process.env.CODEPILOT_FEISHU_APP_ID || '';
const appSecret = process.env.CODEPILOT_FEISHU_APP_SECRET || '';
const pythonCmd = process.env.CODEPILOT_FEISHU_PYTHON || 'python';

if (!appId || !appSecret) {
  console.error('Missing CODEPILOT_FEISHU_APP_ID or CODEPILOT_FEISHU_APP_SECRET');
  process.exit(1);
}

const client = new Lark.Client({
  appId,
  appSecret,
  appType: Lark.AppType.SelfBuild,
  domain: Lark.Domain.Feishu,
});

function log(message, extra) {
  if (typeof extra === 'undefined') {
    console.log(`[feishu-worker] ${message}`);
    return;
  }
  console.log(`[feishu-worker] ${message}`, extra);
}

function extractText(rawContent) {
  if (!rawContent) return '';
  try {
    const parsed = JSON.parse(rawContent);
    if (parsed && typeof parsed.text === 'string') {
      return parsed.text;
    }
  } catch {}
  return String(rawContent);
}

function invokePython(payload) {
  log('dispatch python handler', { chat_id: payload.chat_id, text: payload.text });
  const result = spawnSync(
    pythonCmd,
    ['-m', 'codepilot', 'feishu', 'handle-event'],
    {
      input: JSON.stringify(payload),
      encoding: 'utf8',
      maxBuffer: 1024 * 1024,
      cwd: process.cwd(),
      env: process.env,
    },
  );
  if (result.error) {
    throw result.error;
  }
  const parsed = parsePythonReply(result.stdout || '');
  if (parsed) {
    return parsed;
  }
  if (result.status !== 0) {
    throw new Error((result.stderr || result.stdout || '').trim() || `python exit=${result.status}`);
  }
  throw new Error('python handler did not return valid JSON');
}

function parsePythonReply(stdout) {
  const raw = String(stdout || '').trim();
  if (!raw) return null;
  try {
    return JSON.parse(raw);
  } catch {}
  const lines = raw
    .split(/\r?\n/)
    .map(line => line.trim())
    .filter(Boolean);
  for (let index = lines.length - 1; index >= 0; index -= 1) {
    try {
      return JSON.parse(lines[index]);
    } catch {}
  }
  return null;
}

function buildErrorCard(text) {
  return {
    type: 'interactive',
    card: {
      config: { wide_screen_mode: true },
      header: {
        title: { tag: 'plain_text', content: 'CodePilot 飞书处理失败' },
        template: 'red',
      },
      elements: [
        {
          tag: 'div',
          text: { tag: 'lark_md', content: String(text || 'CodePilot 飞书处理失败。') },
        },
      ],
    },
  };
}

function buildProcessingCard(text) {
  const raw = String(text || '').trim();
  const displayText = raw
    ? `已收到您的消息「${raw.slice(0, 200)}」，正在处理中，请稍候...`
    : '已收到您的消息，正在处理中，请稍候...';
  return {
    type: 'interactive',
    card: {
      config: { wide_screen_mode: true },
      header: {
        title: { tag: 'plain_text', content: 'CodePilot 处理中' },
        template: 'blue',
      },
      elements: [
        {
          tag: 'div',
          text: { tag: 'lark_md', content: displayText },
        },
        {
          tag: 'note',
          elements: [
            { tag: 'plain_text', content: '智能体正在处理中，处理完成后会自动更新此卡片。' },
          ],
        },
      ],
    },
  };
}

async function processIncomingMessage(payload) {
  const chatId = payload.chat_id;
  let processingMsgId = null;
  try {
    // 先发送处理中反馈，让用户立即感知消息已被接收，并记录卡片 message_id
    processingMsgId = await sendReply(chatId, buildProcessingCard(payload.text));
    const reply = invokePython(payload);

    // 处理完成后，优先更新已有的处理中卡片（一条消息从"处理中"变成最终结果）
    if (reply && reply.type !== 'ignore') {
      if (processingMsgId && reply.type === 'interactive' && reply.card) {
        // 卡片回复 → 直接更新处理中卡片为最终结果
        await updateReply(chatId, processingMsgId, reply.card);
      } else if (processingMsgId) {
        // 非卡片回复 → 更新处理中卡片标记完成，再发送实际结果
        const doneCard = buildProcessingCard('处理完成，请查看下方回复。');
        doneCard.card.header.template = 'green';
        await updateReply(chatId, processingMsgId, doneCard.card);
        await sendReply(chatId, reply);
      } else {
        // 没有处理中卡片 message_id（极少见），直接发送结果
        await sendReply(chatId, reply);
      }
    }
    // reply.type === 'ignore'：什么都不做，处理中卡片保留在聊天中
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    log('handler failed', { chatId, detail });
    try {
      if (processingMsgId) {
        // 更新处理中卡片为错误状态
        await updateReply(chatId, processingMsgId, buildErrorCard(`CodePilot 飞书处理失败：${detail}`).card);
      } else {
        await sendReply(chatId, buildErrorCard(`CodePilot 飞书处理失败：${detail}`));
      }
    } catch (replyError) {
      log('failed to send error reply', {
        chatId,
        detail: replyError instanceof Error ? replyError.message : String(replyError),
      });
    }
  }
}

function cardActionEvent(data) {
  const event = data?.event && typeof data.event === 'object' ? data.event : data;
  return event && typeof event === 'object' ? event : {};
}

function cardActionChatId(event) {
  return String(
    event?.chat_id
    || event?.open_chat_id
    || event?.context?.open_chat_id
    || event?.context?.chat_id
    || '',
  );
}

function cardActionMessageId(event) {
  return String(
    event?.message_id
    || event?.open_message_id
    || event?.context?.open_message_id
    || '',
  );
}

function cardActionCommand(event) {
  const value = event?.action?.value;
  if (!value || typeof value !== 'object') return '';
  return String(value.command || value.cmd || value.text || '').trim();
}

async function processCardAction(data) {
  const event = cardActionEvent(data);
  const chatId = cardActionChatId(event);
  const command = cardActionCommand(event);
  if (!chatId || !command) {
    log('ignore card action without command', { chatId, command });
    return { toast: { type: 'warning', content: '这个按钮没有可执行命令' } };
  }
  let processingMsgId = null;
  try {
    processingMsgId = await sendReply(chatId, buildProcessingCard(command));
    const reply = invokePython({
      event_type: 'card.action.trigger',
      text: command,
      chat_id: chatId,
      message_id: cardActionMessageId(event),
      event_id: data?.header?.event_id || data?.event_id || '',
      action: event.action || {},
      context: event.context || {},
      operator: event.operator || {},
    });
    if (reply && reply.type !== 'ignore') {
      if (processingMsgId && reply.type === 'interactive' && reply.card) {
        await updateReply(chatId, processingMsgId, reply.card);
      } else if (processingMsgId) {
        const doneCard = buildProcessingCard('操作已完成。');
        doneCard.card.header.template = 'green';
        await updateReply(chatId, processingMsgId, doneCard.card);
        await sendReply(chatId, reply);
      } else {
        await sendReply(chatId, reply);
      }
    }
    return { toast: { type: 'success', content: '已执行' } };
  } catch (error) {
    const detail = error instanceof Error ? error.message : String(error);
    log('card action failed', { chatId, detail });
    try {
      if (processingMsgId) {
        await updateReply(chatId, processingMsgId, buildErrorCard(`CodePilot 飞书处理失败：${detail}`).card);
      } else {
        await sendReply(chatId, buildErrorCard(`CodePilot 飞书处理失败：${detail}`));
      }
    } catch (replyError) {
      log('failed to send card action error reply', {
        chatId,
        detail: replyError instanceof Error ? replyError.message : String(replyError),
      });
    }
    return { toast: { type: 'error', content: '执行失败' } };
  }
}

function buildPostContent(title, text, content = null) {
  if (Array.isArray(content)) {
    return {
      zh_cn: {
        title: String(title || 'CodePilot 回复').slice(0, 120),
        content,
      },
    };
  }
  const paragraphs = String(text || '')
    .split(/\n+/)
    .map(line => line.trim())
    .filter(Boolean)
    .slice(0, 20)
    .map(line => [{ tag: 'text', text: line }]);
  return {
    zh_cn: {
      title: String(title || 'CodePilot 回复').slice(0, 120),
      content: paragraphs.length ? paragraphs : [[{ tag: 'text', text: 'CodePilot 已收到。' }]],
    },
  };
}

async function sendReply(chatId, reply) {
  if (!reply || reply.type === 'ignore') {
    log('skip reply', { chatId, reason: 'ignore' });
    return null;
  }
  if (reply.type === 'multi' && Array.isArray(reply.messages)) {
    log('send multi reply', { chatId, count: reply.messages.length });
    let lastMsgId = null;
    for (const message of reply.messages) {
      lastMsgId = await sendReply(chatId, message);
    }
    return lastMsgId;
  }
  let res;
  if (reply.type === 'interactive' && reply.card) {
    log('send interactive reply', { chatId });
    res = await client.im.message.create({
      params: { receive_id_type: 'chat_id' },
      data: {
        receive_id: chatId,
        msg_type: 'interactive',
        content: JSON.stringify(reply.card),
      },
    });
    return res?.data?.message_id || null;
  }
  if (reply.type === 'post') {
    const title = String(reply.title || 'CodePilot 详情');
    log('send post reply', { chatId, title: title.slice(0, 80) });
    res = await client.im.message.create({
      params: { receive_id_type: 'chat_id' },
      data: {
        receive_id: chatId,
        msg_type: 'post',
        content: JSON.stringify(buildPostContent(title, reply.text || '', reply.content || null)),
      },
    });
    return res?.data?.message_id || null;
  }
  const text = String(reply.text || 'CodePilot 已收到，但没有可发送的结果。');
  log('send rich text reply', { chatId, preview: text.slice(0, 80) });
  res = await client.im.message.create({
    params: { receive_id_type: 'chat_id' },
    data: {
      receive_id: chatId,
      msg_type: 'post',
      content: JSON.stringify(buildPostContent('CodePilot 回复', text)),
    },
  });
  return res?.data?.message_id || null;
}

async function updateReply(chatId, messageId, card) {
  if (!messageId || !card) {
    log('skip update reply', { chatId, reason: !messageId ? 'no message_id' : 'no card' });
    return;
  }
  log('update interactive reply', { chatId, messageId });
  await client.im.message.update({
    params: { message_id: messageId },
    data: {
      msg_type: 'interactive',
      content: JSON.stringify(card),
    },
  });
}

const dispatcher = new Lark.EventDispatcher({}).register({
  'im.message.receive_v1': (data) => {
    const message = data?.message || {};
    if (String(message.message_type || '').toLowerCase() !== 'text') {
      log('ignore non-text message', { message_type: message.message_type || '' });
      return;
    }
    const chatId = message.chat_id;
    const text = extractText(message.content);
    if (!chatId || !text.trim()) {
      log('ignore empty message', { chatId, text });
      return;
    }
    const eventId = data?.header?.event_id || data?.event_id || '';
    log('received text message', { chatId, event_id: eventId, message_id: message.message_id || '', text });
    void processIncomingMessage({
      text,
      chat_id: chatId,
      message_id: message.message_id || '',
      event_id: eventId,
    });
  },
  'card.action.trigger': async (data) => {
    const event = cardActionEvent(data);
    log('received card action', {
      chat_id: cardActionChatId(event),
      message_id: cardActionMessageId(event),
      command: cardActionCommand(event),
    });
    return processCardAction(data);
  },
});

const wsClient = new Lark.WSClient({
  appId,
  appSecret,
  loggerLevel: Lark.LoggerLevel.info,
});

process.on('uncaughtException', (error) => {
  console.error('[feishu-worker] uncaughtException', error);
});

process.on('unhandledRejection', (error) => {
  console.error('[feishu-worker] unhandledRejection', error);
});

log('starting long connection');
wsClient.start({ eventDispatcher: dispatcher });

File: codepilot/feishu_worker.py
"""Feishu WebSocket long-connection worker — Python native replacement for feishu_worker.mjs.

Uses lark-oapi SDK for WebSocket event subscription and REST API message sending.
All event handling is in-process (no subprocess per message).
"""

from __future__ import annotations

import json
import os
import sys
from typing import Any

from lark_oapi import Client, EventDispatcherHandler, LogLevel
from lark_oapi.api.im.v1 import (
    CreateMessageRequest,
    CreateMessageRequestBody,
    PatchMessageRequest,
    PatchMessageRequestBody,
    P2ImMessageReceiveV1,
)
from lark_oapi.ws import Client as WSClient

from codepilot.feishu_bot import handle_event_payload, handle_card_action_payload
from codepilot.feishu_bot.constants import _NOTIFY_DEDUPE_SERVICE

APP_ID = os.environ.get("CODEPILOT_FEISHU_APP_ID", "")
APP_SECRET = os.environ.get("CODEPILOT_FEISHU_APP_SECRET", "")


def _log(msg: str, extra: Any = None) -> None:
    if extra is not None:
        print(f"[feishu-worker] {msg}", extra)
    else:
        print(f"[feishu-worker] {msg}")


def _extract_text(raw_content: str) -> str:
    if not raw_content:
        return ""
    try:
        parsed = json.loads(raw_content)
        if isinstance(parsed, dict) and isinstance(parsed.get("text"), str):
            return parsed["text"]
    except (json.JSONDecodeError, TypeError):
        pass
    return str(raw_content)


def _build_error_card(text: str) -> dict[str, Any]:
    return {
        "type": "interactive",
        "card": {
            "config": {"wide_screen_mode": True},
            "header": {
                "title": {"tag": "plain_text", "content": "CodePilot 飞书处理失败"},
                "template": "red",
            },
            "elements": [
                {
                    "tag": "div",
                    "text": {"tag": "lark_md", "content": str(text or "CodePilot 飞书处理失败。")},
                },
            ],
        },
    }


def _build_processing_card(text: str) -> dict[str, Any]:
    raw = str(text or "").strip()
    display = (
        f"已收到您的消息「{raw[:200]}」，正在处理中，请稍候..."
        if raw
        else "已收到您的消息，正在处理中，请稍候..."
    )
    return {
        "type": "interactive",
        "card": {
            "config": {"wide_screen_mode": True},
            "header": {
                "title": {"tag": "plain_text", "content": "CodePilot 处理中"},
                "template": "blue",
            },
            "elements": [
                {"tag": "div", "text": {"tag": "lark_md", "content": display}},
                {
                    "tag": "note",
                    "elements": [
                        {"tag": "plain_text", "content": "智能体正在处理中，处理完成后会自动更新此卡片。"},
                    ],
                },
            ],
        },
    }


def _build_post_content(title: str, text: str, content: Any = None) -> dict[str, Any]:
    if isinstance(content, list):
        return {"zh_cn": {"title": str(title)[:120], "content": content}}

    paragraphs = [
        [{"tag": "text", "text": line}]
        for line in str(text or "").split("\n")
        if line.strip()
    ][:20]
    return {
        "zh_cn": {
            "title": str(title or "CodePilot 回复")[:120],
            "content": paragraphs if paragraphs else [[{"tag": "text", "text": "CodePilot 已收到。"}]],
        },
    }


class FeishuWorker:
    def __init__(self, app_id: str, app_secret: str) -> None:
        self._app_id = app_id
        self._app_secret = app_secret
        self._rest_client = (
            Client.builder()
            .app_id(app_id)
            .app_secret(app_secret)
            .log_level(LogLevel.INFO)
            .build()
        )

    # ── REST API helpers ──────────────────────────────────────────────

    def send_reply(self, chat_id: str, reply: dict[str, Any]) -> str | None:
        """Send a reply to a chat. Returns message_id or None."""
        if not reply or reply.get("type") == "ignore":
            _log("skip reply", {"chat_id": chat_id, "reason": "ignore"})
            return None

        if reply.get("type") == "multi" and isinstance(reply.get("messages"), list):
            _log("send multi reply", {"chat_id": chat_id, "count": len(reply["messages"])})
            last_msg_id = None
            for msg in reply["messages"]:
                last_msg_id = self.send_reply(chat_id, msg)
            return last_msg_id

        if reply.get("type") == "interactive" and reply.get("card"):
            _log("send interactive reply", {"chat_id": chat_id})
            req = (
                CreateMessageRequest.builder()
                .request_body(
                    CreateMessageRequestBody.builder()
                    .receive_id(chat_id)
                    .msg_type("interactive")
                    .content(json.dumps(reply["card"], ensure_ascii=False))
                    .build()
                )
                .build()
            )
            resp = self._rest_client.im.v1.message.create(req, lark_oapi_option("receive_id_type", "chat_id"))
            return resp.data.message_id if resp.data else None

        if reply.get("type") == "post":
            title = str(reply.get("title", "CodePilot 详情"))
            _log("send post reply", {"chat_id": chat_id, "title": title[:80]})
            req = (
                CreateMessageRequest.builder()
                .request_body(
                    CreateMessageRequestBody.builder()
                    .receive_id(chat_id)
                    .msg_type("post")
                    .content(
                        json.dumps(
                            _build_post_content(title, reply.get("text", ""), reply.get("content")),
                            ensure_ascii=False,
                        )
                    )
                    .build()
                )
                .build()
            )
            resp = self._rest_client.im.v1.message.create(req, lark_oapi_option("receive_id_type", "chat_id"))
            return resp.data.message_id if resp.data else None

        text = str(reply.get("text", "CodePilot 已收到，但没有可发送的结果。"))
        _log("send rich text reply", {"chat_id": chat_id, "preview": text[:80]})
        req = (
            CreateMessageRequest.builder()
            .request_body(
                CreateMessageRequestBody.builder()
                .receive_id(chat_id)
                .msg_type("post")
                .content(json.dumps(_build_post_content("CodePilot 回复", text), ensure_ascii=False))
                .build()
            )
            .build()
        )
        resp = self._rest_client.im.v1.message.create(req, lark_oapi_option("receive_id_type", "chat_id"))
        return resp.data.message_id if resp.data else None

    def update_reply(self, chat_id: str, message_id: str, card: dict[str, Any]) -> None:
        """Patch an existing interactive card message."""
        if not message_id or not card:
            _log("skip update reply", {"chat_id": chat_id, "reason": "no message_id" if not message_id else "no card"})
            return
        _log("patch interactive reply", {"chat_id": chat_id, "message_id": message_id})
        req = (
            PatchMessageRequest.builder()
            .message_id(message_id)
            .request_body(
                PatchMessageRequestBody.builder()
                .content(json.dumps(card, ensure_ascii=False))
                .build()
            )
            .build()
        )
        self._rest_client.im.v1.message.patch(req)

    # ── Event handlers ────────────────────────────────────────────────

    def _handle_message(self, event: P2ImMessageReceiveV1) -> None:
        data = event.event
        if not data or not data.message:
            return

        message = data.message
        if str(message.message_type or "").lower() != "text":
            _log("ignore non-text message", {"message_type": str(message.message_type or "")})
            return

        chat_id = message.chat_id
        text = _extract_text(message.content or "")
        if not chat_id or not text.strip():
            _log("ignore empty message", {"chat_id": chat_id, "text": text})
            return

        event_id = getattr(event, "event_id", "") or ""
        _log(
            "received text message",
            {"chat_id": chat_id, "event_id": event_id, "message_id": message.message_id or "", "text": text},
        )

        self._process_incoming_message(
            {
                "text": text,
                "chat_id": chat_id,
                "message_id": message.message_id or "",
                "event_id": event_id,
            }
        )

    def _handle_card_action(self, card: Any) -> Any:
        event = card.action or {}
        chat_id = str(
            event.open_chat_id
            or getattr(event, "context", {}).get("open_chat_id", "")
            or getattr(card, "open_chat_id", "")
            or ""
        )
        command_value = ""
        if isinstance(event, dict):
            value = event.get("value") or {}
            if isinstance(value, dict):
                command_value = str(value.get("command") or value.get("cmd") or value.get("text") or "").strip()
        elif hasattr(event, "value") and getattr(event, "value", None):
            v = getattr(event, "value")
            if isinstance(v, dict):
                command_value = str(v.get("command") or v.get("cmd") or v.get("text") or "").strip()

        if not chat_id or not command_value:
            _log("ignore card action without command", {"chat_id": chat_id, "command": command_value})
            return {"toast": {"type": "warning", "content": "这个按钮没有可执行命令"}}

        _log("received card action", {"chat_id": chat_id, "command": command_value})

        try:
            processing_msg_id = self.send_reply(chat_id, _build_processing_card(command_value))
            reply = handle_card_action_payload(
                {
                    "event_type": "card.action.trigger",
                    "text": command_value,
                    "chat_id": chat_id,
                    "message_id": str(getattr(card, "open_message_id", "") or ""),
                    "event_id": getattr(event, "event_id", "") if isinstance(event, dict) else "",
                    "action": event if isinstance(event, dict) else {},
                    "context": getattr(event, "context", {}) if hasattr(event, "context") else {},
                    "operator": getattr(event, "operator", {}) if hasattr(event, "operator") else {},
                }
            )

            if reply and reply.get("type") != "ignore":
                if processing_msg_id and reply.get("type") == "interactive" and reply.get("card"):
                    self.update_reply(chat_id, processing_msg_id, reply["card"])
                elif processing_msg_id:
                    done = _build_processing_card("操作已完成。")
                    done["card"]["header"]["template"] = "green"
                    self.update_reply(chat_id, processing_msg_id, done["card"])
                    self.send_reply(chat_id, reply)
                else:
                    self.send_reply(chat_id, reply)

            return {"toast": {"type": "success", "content": "已执行"}}
        except Exception as exc:
            detail = str(exc)
            _log("card action failed", {"chat_id": chat_id, "detail": detail})
            return {"toast": {"type": "error", "content": "执行失败"}}

    def _process_incoming_message(self, payload: dict[str, Any]) -> None:
        chat_id = payload["chat_id"]
        processing_msg_id = None
        try:
            processing_msg_id = self.send_reply(chat_id, _build_processing_card(payload["text"]))
            reply = handle_event_payload(payload)

            if reply and reply.get("type") != "ignore":
                if processing_msg_id and reply.get("type") == "interactive" and reply.get("card"):
                    self.update_reply(chat_id, processing_msg_id, reply["card"])
                elif processing_msg_id:
                    done = _build_processing_card("处理完成，请查看下方回复。")
                    done["card"]["header"]["template"] = "green"
                    self.update_reply(chat_id, processing_msg_id, done["card"])
                    self.send_reply(chat_id, reply)
                else:
                    self.send_reply(chat_id, reply)
        except Exception as exc:
            detail = str(exc)
            _log("handler failed", {"chat_id": chat_id, "detail": detail})
            try:
                if processing_msg_id:
                    self.update_reply(
                        chat_id,
                        processing_msg_id,
                        _build_error_card(f"CodePilot 飞书处理失败：{detail}")["card"],
                    )
                else:
                    self.send_reply(chat_id, _build_error_card(f"CodePilot 飞书处理失败：{detail}"))
            except Exception as reply_err:
                _log("failed to send error reply", {"chat_id": chat_id, "detail": str(reply_err)})

    # ── Main entry ────────────────────────────────────────────────────

    def run(self) -> None:
        _log("starting long connection")

        event_handler = (
            EventDispatcherHandler.builder(APP_SECRET, "")
            .register_p2_im_message_receive_v1(self._handle_message)
            .register_p2_card_action_trigger(self._handle_card_action)
            .build()
        )

        ws_client = WSClient(
            app_id=self._app_id,
            app_secret=self._app_secret,
            event_handler=event_handler,
            log_level=LogLevel.INFO,
            auto_reconnect=True,
        )

        try:
            ws_client.start()
        except KeyboardInterrupt:
            _log("worker shutting down")
        except Exception as exc:
            _log(f"worker exited with error: {exc}")
            raise


def _lark_oapi_option(key: str, value: str):
    """Build a lark_oapi request option for receive_id_type etc."""
    from lark_oapi import RequestOption

    opt = RequestOption()
    setattr(opt, key, value)
    return opt


lark_oapi_option = lambda k, v: _lark_oapi_option(k, v)


def main() -> None:
    """Entry point for `python -m codepilot feishu run-worker`."""
    if not APP_ID or not APP_SECRET:
        print("Missing CODEPILOT_FEISHU_APP_ID or CODEPILOT_FEISHU_APP_SECRET", file=sys.stderr)
        sys.exit(1)

    worker = FeishuWorker(APP_ID, APP_SECRET)
    worker.run()


if __name__ == "__main__":
    main()
